src/dmd/typesem.d

```diff
@@ -3223,7 +3223,24 @@ private extern(C++) final class DotExpVisitor : Visitor
                 }
                 checkAccess(e.loc, sc, null, v);
                 Expression ve = new VarExp(e.loc, v);
-                ve = ve.expressionSemantic(sc);
+                // In case the expression is something known trivial, like this.enum, super.enum,
+                // variable.enum or type.enum, where we know the lhs does not have side effects,
+                // just use the enum directly.
+                // It would be preferable if we could just directly check if the left hand side of the
+                // expression is obviously side-effect free, but there does not seem to be a way to do that.
+                if (e.op == TOK.type || e.op == TOK.this_ || e.op == TOK.super_ || e.op == TOK.variable)
+                {
+                    ve = ve.expressionSemantic(sc);
+                }
+                else
+                {
```

Comment:
This fix looks like it's in the wrong place. `expressionSemantic()` should *not* be removing expressions with side effects. Putting the fix here is papering over the problem for this case, and it'll just cause other problems. Need to find the root problem.

Reply:
Right, the problem is if I keep the side effects via comma expression, as I have to in order to get correct code in the case call().ENUM, then I end up with unusable comma expressions downstream. I can't fix it earlier, because this is where the access resolves.

Reply:
> what DMD does right now is removing side effects from expressions in that code: call().ENUM becomes ENUM, with call() being discarded.

That's a bug, as call() has side effects. That's where it needs to be fixed.

> The problem is fixing that without breaking enum access for trivial lhs, say alias foo = this.bar.

Maybe that's there the special case should go.

Reply:
I don't understand what you mean. This is where that happens.